### DISCUSSION TITLE
Sustain should sustain notes after their note-off

### DISCUSSION THREAD

**Discussion by ybungalobill** [FluidSynth/fluidsynth#1214] [Ideas]
On a real piano, if one releases a key and then quickly depresses the sustain pedal, the remaining vibrations from that note will be sustained. A midi keyboard I have simulates that behavior. However, with fluidsynth the note keeps attenuating quickly just as if the sustain wasn't pressed.

I understand that a fully realistic sustain behavior (resonances, etc) might be out-of-scope for fluidsynth. However, this particular aspect of the way sustain works should be rather simple to introduce.

Is this something that we may see implemented in fluidsynth? Should I open an issue?

Here's somebody else ran into the same issue: https://linuxmusicians.com/viewtopic.php?t=18829

**Comment by derselbst** [comment]
Section 9.2 of the SoundFont2 spec says:

> MIDI CC64 Sustain - ACTIVE when greater than or equal to 64. When the sustain function is active, all notes in the keyon state remain in the key-on state regardless of whether a key-off command for the note arrives. The key-off commands
are stored, and when sustain becomes inactive, all stored key-off commands are executed. 

And this is exactly how it's implemented in fluidsynth. The "key-off" immediately enters the release phase of the AHDSR envelope - there is no way of "going back" in the envelope. This is also the behavior I perceive from my EMU synths.

I do perceive the behavior you've described from my Roland ePiano - however, only for the piano based sounds (Harpsichord, Piano, ePiano, Vibraphone, etc.), but not for strings or pipe organs.

> However, this particular aspect of the way sustain works should be rather simple to introduce.

I don't see a simple solution for this, as the solution must fit into the SF2 synth model and deal with all kinds of different tunes and not just a piano. Yet, if you have a simple idea, you're welcome to prove me wrong and we might consider implementing it.

**Reply by ybungalobill** [reply]
I'm not familiar with the SoundFont spec or the architecture of fluidsynth to say exactly how to implement it, or how to distinguish which instruments to apply it to. However, conceptually, when a sustain becomes active, the notes in their release phase should go back to their decay part of the envelope -- with amplitude adjusted, obviously.

P.S. for reference: I observe this behavior with a Kawai keyboard. In general, I'd expect that to be the case with any but the lowest end instruments.

**Reply by ybungalobill** [reply]
I think the following logic would do the trick to distinguish piano from other instruments:

When sustain becomes ON, for notes in the release phase:
* if the envelope is above the Sustain level -- go back to decay.
* otherwise -- keep it in release.

To my understanding organ and strings have sustain levels close to 1, so even after a short time in release they'll drop below the threshold and won't be picked up by a sustain. With a piano-like instruments the sustain level is 0, so after short time in release the value will still be above the sustain level, and it will be picked up by a repeated sustain-on event.

**Reply by derselbst** [reply]
Ok, thank you. That sounds reasonable. Yet, I currently don't see a simple way of implementing this behavior in fluidsynth. But I would welcome anybody who proposes a solution or PR to this.

**Reply by RobsonFBP** [reply]
Some time ago (3 years) I also raised this question.
these days I've been looking at the fluidsynth source code, it seems possible to make the sustain pedal work as a soft pedal without violating the soundfont specification, adding a variable in the voice structure that would only be triggered if a specific modulator was added: source1 CC64, destination volume envelope release.

**Comment by RobsonFBP** [comment]
Sustain should sustain notes after notes off

**Comment by midimusic** [comment]
I have just tried to use fluidsynth's sustain with an sf2 piano sound font. Basically fluidsynth's sustain  DOES NOTHING as the notes behave and sound exactly the same whether sustain is on or off.
As RobsonFBP says above a sustained note MUST continue to sound AFTER the key is released. With fluidsynth sustained notes finish their sustain phase when the key is released.
Sustain can easily be implemented by delaying "note off" events until the sustain is turned off. I have tried this and it does correctly simulate a piano sustain pedal. It is even possible to play a note multiple times while sustain is on and get the correct sound. (When sustain is released all the delayed "note off" events must be sent).
csw900
